### PR TITLE
docs: align guidance files for humans and agents

### PR DESCRIPTION
**Pull request by keulinho**
### 1. Why is this change necessary?

Repository guidance is currently split between README files, AGENTS files, local overrides, and tool-specific entry points. That makes it easy for human-facing and agent-facing instructions to drift.

### 2. What does this change do, exactly?

- Documents the guidance-file model in the root `AGENTS.md` and a new ADR.
- Moves reusable normative Administration guidance into `coding-guidelines/administration/`.
- Keeps folder-specific working guidance in README files and makes AGENTS files route to those READMEs.
- Adds `CLAUDE.md` and `GEMINI.md` files next to every `AGENTS.md`, importing the local AGENTS file with `@AGENTS.md`.
- Links relevant READMEs, coding guidelines, and ADRs where that adds context.

### 3. Describe each step to reproduce the issue or behaviour.

1. Open a subtree with working guidance.
2. Read the local README as the source of truth.
3. Open the local `AGENTS.md`, `CLAUDE.md`, or `GEMINI.md` and confirm it routes to the same guidance instead of duplicating it.

### 4. Please link to the relevant issues (if any).

None.

### 5. Checklist

- [ ] I have written tests and verified that they fail without my change
- [ ] I have updated developer-facing release notes if this change is **relevant** for external developers:
  - Add a short entry to `RELEASE_INFO-6.<major>.md` under “Upcoming” for informational changes, including the consequences of the change and how it affects external developers.
  - Add an `UPGRADE` section in `UPGRADE-6.<next-major>.md` for breaking changes (what/why/impact/how to adapt).
  - See the [Documenting a Release Process](https://github.com/shopware/shopware/blob/trunk/delivery-process/documenting-a-release.md) for details.
- [x] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [ ] I have read the contribution requirements and fulfilled them

Checked locally: `git diff --check`; verified every `AGENTS.md` has matching `CLAUDE.md` and `GEMINI.md` import files.

### CODE REVIEW

File: AGENTS.md [hunk 1]
```diff
@@ -50,6 +50,127 @@ Skills can have an optional unattended twin via [GitHub Agentic Workflows](https
...
+- Do not duplicate ADR or coding-guideline content in READMEs. Summarize the local implication and link to the source.
+- Reserve the root `AGENTS.md` for short repo-wide instructions that agents must see before working; move detailed or component-specific guidance into the relevant README or coding guideline.
+- Folders whose README contains working guidance should have an `AGENTS.md` stub that points agents to the README. Do not duplicate README content in that stub.
+- Standalone overview/index READMEs do not need an `AGENTS.md` stub. This includes the main component READMEs under `src/*` and folders whose path contains `docs`.
+- Do not add README or AGENTS files just to index `coding-guidelines/`. Component READMEs should link directly to the concrete guideline files that apply.
+- Keep agent-only mechanics in `AGENTS.md` only when they are not useful to humans. The repository root `AGENTS.md` is the exception because it carries global agent routing.
+
+## Administration ACL
```
Comment (keulinho) [inline AGENTS.md]: basically everything that is following here was the rules i accumulated over the last weeks in my local agents.override

by design they contain opiinioated guidelines, but in my experience it makde it so much more effective to work with agents

File: adr/2026-06-17-shared-guidance-files-for-humans-and-agents.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
+---
+title: Shared Guidance Files for Humans and Agents
+date: 2026-06-17
+area: process
+tags: [documentation, agents, developer-experience, coding-guidelines, adr]
+status: accepted
+---
+
+## Context
```
Comment (keulinho) [inline adr/2026-06-17-shared-guidance-files-for-humans-and-agents.md:9]: read this and the agents.md changes first

the other changes based on that decision are then basically mechanically and logically following from the decision described here and the guideline in the agent.md file

File: coding-guidelines/core/unit-tests.md [hunk 3]
```diff
@@ -48,10 +48,12 @@ More broadly speaking, it is hard to guarantee that the mock behaves in the same
 Use mocks only where you need to because:
 1. creating the objects is hard as you need tons of nested dependencies to create the object.
 or
-2. the class produces some side effects that you don't want in unit tests (e.g., DB writes).
+2. the class produces some side effects that you don't want in unit tests.
 
 For all other cases, use real implementations and rely as minimally as possible on the magic of phpunit's mocking framework.
 
+Do not mock Doctrine DBAL `Connection` in unit tests. Classes with direct SQL/DBAL work should usually be treated as database adapters and covered with integration tests. Unit-test the surrounding business logic against a narrow repository, gateway, or service abstraction instead.
+
```
Comment (nfortier-shopware) [inline coding-guidelines/core/unit-tests.md:57]: But use stubs, createStub() to disable it when consumed by objects you are testing.

To clear any ambiguity, using a mock for DBAL with assertions (such expects with) brings not much value especially when it gets super conditional with raw SQL. While a stub that can return a result object is more interesting for the rest of your set.

More concrete stubs like StaticEntityRepository are also handy

File: coding-guidelines/core/unit-tests.md [hunk 4]
```diff
@@ -8,7 +8,7 @@ When writing unit tests, the following is important:
 
 - **"100% coverage "** - This does not mean that simply a high code coverage should be generated, but that all use cases of each individual service is tested.
 - **Performance** - As we grow more and more it is advisable to pay attention to the speed of the tests.
-- **Mocking** - Don't be lazy but deal with mock objects to optimize for example database access. So you don't have to persist every storage case before but you can describe it as a Mock.
+- **Mocking** - Use mocks intentionally to keep unit tests fast and focused. Mock narrow application abstractions when needed, but do not mock Doctrine DBAL `Connection` directly; isolate SQL/DBAL work in database adapters and cover those adapters with integration tests.
```
Comment (nfortier-shopware) [inline coding-guidelines/core/unit-tests.md]: About coverage for source files, use `@codeCoverageIgnore` annotation with `@see` with the integration test covering it

(it's in the general AGENTS.md)

File: AGENTS.md [hunk 5]
```diff
@@ -50,6 +50,127 @@ Skills can have an optional unattended twin via [GitHub Agentic Workflows](https
...
+
+- Admin UI changes that read or persist DAL entities or associations must update matching ACL privilege mappings and migrations for existing roles when needed; see `coding-guidelines/administration/architecture.md`.
+
+## PHP Code Structure
+
+- Keep application/domain services hexagonal: controllers, CLI commands, subscribers, and handlers translate infrastructure details (`Request`, IO, database, filesystem, HTTP) into plain inputs before calling services.
+- Services must not perform direct infrastructure work or depend on framework objects. Depend on narrow abstractions instead, such as repositories, filesystem interfaces, HTTP clients, or gateways.
+- Services must be unit-testable without external systems; test infrastructure adapters with integration tests.
+- Mark infrastructure adapters `@internal` by default.
+- Mark services `@private` by default when third-party code may call them but must not extend, decorate, or rely on their internals. Use explicit public extension points for supported customization.
```
Comment (mitelg) [inline AGENTS.md]: no 🫣 not a single PHP file in our project has this annotation 😬

File: AGENTS.md [hunk 6]
```diff
@@ -50,6 +50,127 @@ Skills can have an optional unattended twin via [GitHub Agentic Workflows](https
...
+- Do not add `@final` to classes already marked `@internal`; the internal marker is enough for implementation details.
+- Do not repeat `@internal` on constructors or methods inside an `@internal` class.
+- Prefer existing Shopware extension mechanisms over new provider interfaces when they already express the contract, for example Twig inheritance, DAL entities, Admin API routes, or explicit Twig blocks.
+- For new feature designs, explicitly separate the BC-promised public surface from internal implementation services. Document public REST/Admin/Store API contracts, DAL entities, template context, and supported extension points; mark controllers, subscribers, loaders, renderers, and discovery services `@internal` unless they are intended extension points.
+- Be conservative with DTOs/value objects. Add one only when it expresses a meaningful domain concept, crosses a real boundary, or simplifies a public contract. Prefer scalars or arrays for simple internal data, and do not create DTOs solely to model private handoffs inside one class.
```
Comment (mitelg) [inline AGENTS.md]: I would like to argue against that! imo, array as typehint should only be used, if you have a list of objects. associative arrays should not be allowed at all!

File: AGENTS.md [hunk 7]
```diff
@@ -50,6 +50,127 @@ Skills can have an optional unattended twin via [GitHub Agentic Workflows](https
...
+- For unit tests around file access, choose the lightest setup that still reads naturally: simple single-file reads/writes can use Symfony `Filesystem` injected into the class and mocked in the test; when the scenario needs several consecutive filesystem calls, realistic paths, or directory structure, prefer committed `_fixtures` over building temp files at runtime or over-mocking the filesystem.
+- Keep test helpers smaller than the code they replace. Do not hide assertions or feature-flag toggling behind abstractions when direct assertions are just as readable.
+- Prefer one focused test method per distinct exception or behavior over broad data providers when each case has its own meaning.
+- In PHPUnit tests, prefer `expectExceptionObject()` over `expectExceptionMessage()` / `expectExceptionMessageMatches()` because message expectations are deprecated in future PHPUnit versions. Build the expected exception through the same domain factory when one exists so class, code, and message stay aligned with production behavior.
```
Comment (mitelg) [inline AGENTS.md]: we have a PHPStan rule for that already, so you can rephrase to only using the domain factory methods

File: AGENTS.md [hunk 8]
```diff
@@ -50,6 +50,127 @@ Skills can have an optional unattended twin via [GitHub Agentic Workflows](https
...
+- Keep legacy feature-flag behavior in dedicated tests that are easy to remove when the flag is removed.
+- In unit tests, current major feature flags are active by default. Test legacy/off behavior by disabling the flag with the `#[DisabledFeatures]` attribute; do not use `Feature::fake()` just to activate the current major flag.
```
Comment (mitelg) [inline AGENTS.md]: that is already stated here `coding-guidelines/core/feature-flags.md` 

so can be removed here, right?

File: AGENTS.md [hunk 9]
```diff
@@ -50,6 +50,127 @@ Skills can have an optional unattended twin via [GitHub Agentic Workflows](https
...
+- In integration tests, the suite may run multiple times with feature flags on and off. Do not use `#[DisabledFeatures]` there for simple legacy/current branching; skip tests explicitly with `Feature::skipTestIfActive()` or `Feature::skipTestIfInActive()` when the current feature-flag value is not the one the scenario expects.
```
Comment (mitelg) [inline AGENTS.md]: already stated here `coding-guidelines/core/unit-tests.md`

and probably more of the stuff mentioned below

File: AGENTS.md [hunk 10]
```diff
@@ -50,6 +50,127 @@ Skills can have an optional unattended twin via [GitHub Agentic Workflows](https
...
+
+- Use named `yield` cases in unit-test data providers instead of returning arrays, even for small providers. This keeps cases readable and avoids materializing large arrays as providers grow.
+- Do not use `yield from` with an inline array for providers. Prefer one explicit `yield 'human readable case description' => [...]` per scenario.
+- Provider case names should explain the scenario and expected behavior, not mechanically restate raw input values. Good names mention the rule being proven, such as priority, normalization, timezone conversion, or boundary handling.
+- Be conservative when deleting "duplicate" provider cases. Remove only exact semantic duplicates that add no coverage, and keep similar-looking cases when they cover distinct edge behavior.
+
+## Migration Structure
+
+- Use the exact current Unix timestamp for new migration class names, file names, and `getCreationTimestamp()` values. Do not use placeholder or rounded timestamps.
+- Do not add tests for empty/no-op `updateDestructive()` implementations; cover meaningful migration behavior in `update()` or destructive migrations that actually change state.
```
Comment (mitelg) [inline AGENTS.md]: also stated already in the coding guidelines. 

should we really make the root Agents.md so big? or should we just move everything into the coding guidelines? 🤔

File: coding-guidelines/core/unit-tests.md (hunk 4 above)
Comment (keulinho) [reply coding-guidelines/core/unit-tests.md]: good catch, this guideline file precedet the coverageIgnore handling, but i've adjusted it

i let the AI have a single pass to compare coding guidelines with conflicting rules in agents.md, but it didn't report this

File: src/Core/Framework/Api/ApiDefinition/README.md [hunk 11]
```diff
@@ -2,6 +2,10 @@
 
 This directory contains the necessary components and schema files for generating the OpenAPI specifications for the StoreAPI and the AdminAPI.
 
+## Related ADRs
+
+- [API version removal](../../../../../adr/2020-12-02-removing-api-version.md)
```
Comment (mitelg) [inline src/Core/Framework/Api/ApiDefinition/README.md]: you linked several very old ADRs on such files. are they really relevant to know for the LLM? 🤔 like we removed Psalm or this one?

File: AGENTS.md [hunk 12]
```diff
@@ -50,6 +50,127 @@ Skills can have an optional unattended twin via [GitHub Agentic Workflows](https
...
+- Folders whose README contains working guidance should have an `AGENTS.md` stub that points agents to the README. Do not duplicate README content in that stub.
+- Standalone overview/index READMEs do not need an `AGENTS.md` stub. This includes the main component READMEs under `src/*` and folders whose path contains `docs`.
+- Do not add README or AGENTS files just to index `coding-guidelines/`. Component READMEs should link directly to the concrete guideline files that apply.
+- Keep agent-only mechanics in `AGENTS.md` only when they are not useful to humans. The repository root `AGENTS.md` is the exception because it carries global agent routing.
+
+## Administration ACL
+
+- Admin UI changes that read or persist DAL entities or associations must update matching ACL privilege mappings and migrations for existing roles when needed; see `coding-guidelines/administration/architecture.md`.
+
+## PHP Code Structure
```
Comment (mitelg) [inline AGENTS.md]: tbh, I would like to put most of this into PHPStan rules, tests, rector rules, you name it 😬 

I know it is easier to write something for the agents, but having a real guardrail is much better, imo. 

At least we should create tasks for us, to reduce this by adding more real guardrails

**Review by mitelg — CHANGES_REQUESTED** [review]

File: AGENTS.md (hunk 9 above)
Comment (h1k3r) [reply AGENTS.md]: By the way, good point. Unit test guidelines should be separate file, not main, when agent does changes not-related to the unit tests - that does not pollute context.

File: AGENTS.md (hunk 7 above)
Comment (nfortier-shopware) [reply AGENTS.md]: Yes for this one, it is better to stay generic. 

Also, sometimes the agent is trying to bypass the rule by using just `expectException` alone, which is still acceptable PHPUnit API and not to be deprecated. The rule cannot be that strict to ban it either, as some potential edge case might occur when we won't match an object.

```
Prefer `expectExceptionObject()` over a broader `expectException`, build the expected exception through the same domain factory when one exists so class, code, and message stay aligned with production behavior.
```

maybe?

File: coding-guidelines/administration/architecture.md [hunk 13]
```diff
@@ -0,0 +1,37 @@
+# Administration architecture
+
+These rules apply to code under `src/Administration/Resources/app/administration`.
+
+## Layers
+
+- Follow existing Administration code style and component patterns in the area you touch.
+- Do not introduce breaking changes to public Administration APIs or extension points without prior discussion.
+- Use the repository-root composer wrappers for Administration linting, formatting, tests, and builds.
+- Keep the dependency direction `module -> app -> core`. Core code must not import from app or modules.
```
Comment (jleifeld) [inline coding-guidelines/administration/architecture.md]: Not quite right.
`core` => Shouldn't contain any Vue related code

But it is allowed that `module` import something from `core`

File: coding-guidelines/administration/architecture.md [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+- Keep module code independent. Do not import directly from another module; communicate through registered services, repositories, routes, stores, or shared app/core code.
+- Keep the boot order `init-pre/ -> init/ -> init-post/` when changing startup code.
+
+## Extension-aware access
+
+- Access factories, services, stores, and components through the global `Shopware` APIs, for example `Shopware.Component`, `Shopware.Service()`, and `Shopware.Store`.
```
Comment (jleifeld) [inline coding-guidelines/administration/architecture.md]: Not 100% true. There are many cases where people use `inject` in components to get the services, etc. Also in boot related files you don't have directly access to everything. There you can't also use it

File: coding-guidelines/administration/architecture.md [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+- Do not import factory internals directly when an extension-aware global API exists.
+- Register global components through the component factory. Local component imports bypass the extension system.
```
Comment (jleifeld) [inline coding-guidelines/administration/architecture.md]: Not 100% true anymore. Only true for components with OptionsAPI and Twig blocks. 

New components written as Vue SFC components with the new extension system and native Blocks can be registered natively

File: AGENTS.md [hunk 16]
```diff
@@ -50,6 +50,127 @@ Skills can have an optional unattended twin via [GitHub Agentic Workflows](https
...
+- If both REST/Admin/Store API contracts and PHP-level APIs or extension points are affected, document them as separate entries in the relevant sections, for example API for REST routes and Core for services, interfaces, abstract classes, decorators, or extension points.
+- In both release notes and upgrade guides, write from the perspective of extension authors, API consumers, operators, or other outside users. Include whether adjacent APIs remain unchanged when that distinction prevents migration mistakes.
+
+## GitHub PR Handling
+
+- Follow the repository pull request template closely, and do not add extra sections that are not in the template.
+- Do not add a separate validation section to Shopware PR descriptions.
+- Keep PR titles in proper conventional commit style when requested, for example `fix: allow TestBootstrapper to activate Composer plugins`.
+- When updating a pull request after review feedback or CI failures, create a new follow up commit explaining what you fixed in that specific commit instead of amending or force-pushing the existing commit. This preserves the review history and allows reviewers to see the changes clearly.
+
```
Comment (jleifeld) [inline AGENTS.md:113]: The initial  AGENTS.md file is getting very big. But from my experience we should try to avoid filling up the context with too much information at the beginning.

Example: If I give the AI an instruction to do something in an admin module, it doesn't need to know about details of the Core, Storefront, the core or app folder in the admin, etc.

I would prefer something like a codemap. So if the AI touches file xyz. it reads the necessary AGENTS file for it

**Review by jleifeld — CHANGES_REQUESTED** [review]

File: coding-guidelines/administration/architecture.md
Comment (jleifeld) [inline coding-guidelines/administration/architecture.md:1]: The coding guidelines are explaining similar things we already have in our `technical-docs`. I think it is enough to reference them.

Also these guidelines are outside the `administration` subfolder. If people open the AI agent in the subfolder they will never see the coding guidelines. So if we want them, we should move them to the direct `administration` folder

File: AGENTS.md (hunk 6 above)
Comment (keulinho) [reply AGENTS.md]: Please not i remember why i added this rule:
AI came up with a class where it was one public method and internally a few private methods that called each other, codex wanted it to built it cleanly and came up with 5 additionally DTO classes that is absolutely not necessary and most of the time blows up the private methods as well as suddenly they basically only handle DTO transformations

what's the issue with using associative arrays with the proper type hints inside a single class?

File: AGENTS.md (hunk 8 above)
Comment (keulinho) [reply AGENTS.md]: no my agent always worked around this if not mentioned explicitly

File: AGENTS.md (hunk 12 above)
Comment (keulinho) [reply AGENTS.md]: sure this is not intending to replace such guardrails, however some things are basically impossible to verify in a deterministic manner, and if so it is way more complicated

adding a new design rule here takes 10 seconds, building a real guardrail that is enforced in CI might easily take multiple days

Also currently it is intentional that we copy stuff that is already enforced by phpstan rules or similiar stuff, similiar because it's way faster when the AI doesn't need run into the wall and then fix

also for humans it is not nice to reverse engineer the intent behind a phpstan rule, and IMHO it is so much easier to digest when there is a single file listing the most relevant principles/guidelines directly instead of having to go through a bunch of separate files, rules etc to see how stuff should be done

File: src/Core/Framework/Api/ApiDefinition/README.md (hunk 11 above)
Comment (keulinho) [reply src/Core/Framework/Api/ApiDefinition/README.md]: done, in general i think it makes sense to link to those where it makes sense, but here it clearly doesn't add any value

File: AGENTS.md (hunk 5 above)
Comment (keulinho) [reply AGENTS.md]: somehow that sentence did hide there, i only wanted to clarify internal and final annotations, will remove

File: coding-guidelines/administration/architecture.md
Comment (keulinho) [reply coding-guidelines/administration/architecture.md:1]: IMHO up for discussion, i think it's weird that we have multiple locations where we document internal coding guidelines and i wanted to unify that

To be honest the AI created all the admin coding guides based on what is in technical docs already, however my intention was to split it and keep docs in the docs, but put rules and guidelines into the coding guidelines folder

Comment (keulinho) [reply coding-guidelines/administration/architecture.md:1]: > If people open the AI agent in the subfolder they will never see the coding guidelines.

BTW i don't know if that is something we should suggest, as the root agents file contains more, also specifically about release notes, PR handling and so on

File: AGENTS.md (hunk 6 above)
Comment (mitelg) [reply AGENTS.md]: I remember a case, where I refactored a class that only internally used a associative array to a DTO. before it was a PHPStan mess, because the array could have keys or not, but also null values or not. then with the object the handling was much easier, but the code reads almost the same. See https://github.com/shopware/shopware/pull/13253/changes

the code is now much cleaner and more predictable. no PHPStan crutches necessary. 
arrays might be okay, if you are always passing around a predefined dataset, like a DB result, which is always the same. but once you are working with it, and it gets dynamic, it will fuck off sooner or later.

File: AGENTS.md (hunk 8 above)
Comment (mitelg) [reply AGENTS.md]: your goal was to reduce duplication, which is not achieved then :slightly_smiling_face: 
I cannot give help on the technical part of how agents are behaving, but then we should think of removing the coding guidelines maybe? :thinking:

**Review by mitelg — APPROVED** [review]

**Review by nfortier-shopware — APPROVED** [review]

**Review by taltholtmann — CHANGES_REQUESTED** [review]
Nice PR, i like the overall structure, one point I would also align with the other reviewrs is, the big `agends.md`. I think we should be stricter about keeping the root `AGENTS.md` small.

Since the root guidance is always part of the agent startup context, it should probably contain only repo-wide invariants and routing to the relevant subtree guidance. Some sections currently in root, such as `Boyscouting Scope`, `Bug Fix Root Cause And Scope`, `Release Notes And Changelog`, and `GitHub PR Handling`, are useful but seem workflow-specific rather than always-needed context.

Could we move those into more targeted places, for example coding guidelines, README sections, or workflow/skill-specific guidance, and keep root `AGENTS.md` focused on minimal global rules plus routing?

I also wonder if we can verify which files are actually loaded by the supported tools from representative entry points:

- Claude Code from the repository root and from a nested subtree
- Gemini CLI from the repository root and from a nested subtree
- Codex/opencode from the same locations, if applicable

That would make the context-size discussion less speculative. In particular, the broad `CLAUDE.md` / `GEMINI.md` wrapper rollout may behave differently per tool: `@AGENTS.md` imports are not just links, and if a tool loads multiple hierarchical guidance files, many tiny wrappers can still add avoidable context noise.

File: AGENTS.md (hunk 16 above)
Comment (jleifeld) [reply AGENTS.md:113]: A more detailed example:

I tell the AI: "Fix xyz". Then with this AGENTS.md files it reads things about how to create a Github PR, etc. This is not necessary for the task and leads to more hallucination and fills up the context with information not relevant for the task.

A better approach would be, that it just references the contribution guidelines. If you then ask the AI to create a PR description for it, it would read the `CONTRIBUTING.md` file and knows what to do.

So I would look for every aspect in the AGENTS.md: is this really necessary for EVERY task you give to the AI? If not, it should not be inside the AGENTS file

**Review by taltholtmann — APPROVED** [review]

File: AGENTS.md (hunk 16 above)
Comment (keulinho) [reply AGENTS.md:113]: the thing that i noticed is that when stuff is included conditionally it leads to it not being loaded a lot of times when it should

therefore i prefer to tend to have too much in the context, then to miss stuff

but as i said already i'll split a bit further and check more

but one thing: the agents.md currently contains link to the coding guidelines that those are mandatory to be followed, but codex **never** picked that up unless mentioned explicitly

File: AGENTS.md (hunk 6 above)
Comment (keulinho) [reply AGENTS.md]: there can definetly be examples where it totally makes sense, but isn't that exactly what this sentence is trying to convey? 

a blanket rule: never use assoc arrays, or never use DTOs internally will never work, it always depends

so that's why we define a direction here, the tendency i see is that AI tends to overuse DTOs that's why it's phrased the way it is

**Review by socrec — APPROVED** [review]

**Comment by SpiGAndromeda** [comment]
Before this is merged, I'd like to propose one change to the placement rule: key it on how the agent consumes a fact, not on which file is declared the source of truth.

Concretely, anything the agent must have in hand to act (symbol and path maps, hard constraints, invariants, "do X here") belongs in the nearest auto-loaded AGENTS.md and never behind a prose reference. Anything it needs only to understand the module (rationale, the "why", onboarding) belongs in the README, which is safe to read on demand because missing it doesn't break an edit. Reusable normative rules stay in coding-guidelines/, durable decisions in the ADR. One home per fact, cross-link instead of restating.

The driver is maintenance, not context size. As written, decisions 1, 3 and 5 still allow the same rule to live in a README, a subtree AGENTS, or coding-guidelines, which is the duplication this ADR is meant to remove and the gap @mitelg pointed at when he said the dedup goal isn't met. A single mechanical test for where a fact lives is what makes "no duplication" reviewable instead of aspirational.

It also resolves the root-size thread. Keeping root minimal matters because workflow content there loads on every unrelated task; not pushing everything behind references matters because, as @keulinho noted, references often aren't loaded when they should be. Splitting by act-versus-understand removes that conflict: the only things allowed behind an on-demand reference are the ones that don't block an edit if they're missed.

This isn't the "AGENTS as source of truth" option the ADR rejects. AGENTS stays tracked and readable and every human-facing rationale stays in the README, so nothing is hidden from humans.

The measured context-degradation work (position effects, and length effects shown on coding tasks) supports keeping the always-loaded layer small, but I'd treat it as corroboration, not the basis. What would actually sharpen the act/understand line is verifying per tool, from both root and a nested folder, which files are auto-loaded versus only referenced.

The rule in short, by what a fact is for:

- Acting on code (symbol/path maps, constraints, invariants, "do X here") → nearest auto-loaded AGENTS.md, never behind a reference
- Understanding the module (rationale, the "why", onboarding) → README, read on demand
- Reusable normative rules → coding-guidelines/
- Durable decisions and trade-offs → ADR
- One home per fact; cross-link, never restate

**Comment by keulinho** [comment]
Closing this in favor of the new approach by moving it into skills: https://github.com/shopware/shopware/pull/17657
